content/posts/how-to-evolve-application-add-quality/index.md

```diff
@@ -0,0 +1,139 @@
++++
+title = "How to evolve an application to make it more robust and maintainable - Add quality "
+description = ""
+date = 2023-05-29
+featured = false
+draft = false
+comment = true
+toc = true
+reward = true
+pinned = false
+categories = [
+"quality", "test", "CICD", "Sonarqube", "Owasp", "Gradle"
+]
+tags = [
+""
+]
+series = [
+" How to evolve an application to make it more robust and maintainable"
+]
+images = []
++++
+
+<!--more-->
+
+
+## Foreword
+
+In our last articles we looked at how to add tests and how to run them automatically via the CI/CD (GitHub Actions).
+In this article we'll continue to improve the quality of our project using Sonarqube.
+
+
+
+## What is Sonarcloud or Sonarqube
+[index.md](..%2FHowTo%2Fhow-to-use-sonarcloud%2Findex.md)
+
+
+## Integration in java project with Gradle
+
+### Déclare  plugin
+* Add plugin
+
+````groovy 
+plugins {
+id("org.sonarqube") version "4.1.0.3113"
+}
+
+````
+* Add configuration 
+````groovy
+sonar {
+    properties {
+        property "sonar.projectkey", "happraisal"
+        property "sonar.projectName", "happraisal"
+        property "sonar.host.url", "http://localhost:9000"
+        property "sonar.coverage.jacoco.xmlReportPaths", "build/reports/jacoco/all-tests/jacocoAllTestReport.xml"
+        property "sonar.sources", "src/main,build.gradle,gradle"
+        property "sonar.tests", "src/test,src/integration-test,src/system-test"
+        property "sonar.junit.reportPaths", "build/test-results/test,build/test-results/integrationTest,build/test-results/systemTest"
+    }
+}
+````
+
+#### Export results
+
+From intelliJ, via configuration or terminal you can execute :
+```` 
+clean build jacocoTestReport sonar -Dsonar.login=<<YOUR_ID>> -Dsonar.password=<<YOU_PASSWORD>>
+
+````
+
+If you have created a token you can run 
+
+
+````
+./gradlew clean build jacocoTestReport sonar -Dsonar.login=<<TOUR_TOKEN>>
+
+````
+
+### Multi module
+In project with multi-module each module must execute sonar task.
+
+
+## Tools around Sonar
+### Sonar API
+
+Sonar provides some apis to manage project, quality profile, quality gate and others. 
+
+Details can be find [here](https://next.sonarqube.com/sonarqube/web_api/)
+You can use tools like postman or curl to use these apis, but the first thing that you must do is to create a token.
+
+To create a token, after you login  go to your profil and in the security tab and create a token.
+
+For some apis you must create a token with administrator authorizations like 
+* /api/qualitygates/select
+* /api/projects/search
+
+
+
+### Sonarlint
+Sonarlint is a tool that can be used to analyze your code. 
```

Comment:
What kind of analisys?

Reply:
Analyse static : analyse effectuée sans exécuter le code.
On détecte: 
- le code smell (to define)
- les erreure de programmation comme les null pointer exception (compilateur ou ide ?)
- duplication de code



La couverture de code ou les alertes i=owasp doivent être executer en local ou dan s la CI et  les rapports envoyés vers SonarCloud ou Sonarqube